flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -365,4 +401,19 @@ public String toString() {
             return "EmptyDiscardStateObject{" + stateHandleID + '}';
         }
     }
+
+    private boolean preventsDiscardingCreatedCheckpoint(SharedStateEntry entry) {
+        // explicitly set by the backend, e.g. private state is reused
+        if (entry.preventDiscardingCreatedCheckpoint
+                && restoredCheckpointSharingStrategies.containsKey(entry.createdByCheckpointID)) {
+            return true;
+        }
+        // with NO_SHARING strategy, shared state, if any, is bundled inside the checkpoint folder
+        // noinspection RedundantIfStatement
+        if (NO_SHARING == restoredCheckpointSharingStrategies.get(entry.createdByCheckpointID)) {
```

Comment:
IIUC, NO_SHARING means the cp will not be used by next cp, so it's no need to delete it lazily if `NO_SHARING` works as its comments, right ?
Actually, I think there are an issue about NO_SHARING currently. While restoring from `NO_SHARING`, the `materializedSstFiles` should be empty, but currently not as you could see in `RocksDBKeyedStateBackendBuilder#build`.
IIUC, After fixing this, it should be FORMARD not NO_SHARING, right ?
Or I missed something ?
WDYT?

Reply:
Yes, the behavior described in the comment for `NO_SHARING` is relevant only until recovery. 
After the recovery, the state can be re-used.

I don't see any issue with that though; preventing re-use after recovery would result in slower first checkpoint.

Reply:
Okay, Thanks for the clarification.
You mean that `SharingFilesStrategy` only applys on runtime but not on recovery ?
If so, why we need to clarify `NO_SHARING` here,  Do other strategies also need to contain these logics ?

Reply:
> You mean that SharingFilesStrategy only applys on runtime but not on recovery ?

Yes.

> why we need to clarify NO_SHARING here, Do other strategies also need to contain these logics ?

Testing for `NO_SHARING` here allows to delay discarding the checkpoint folder. Delay is necdesary if the checkpoint that was created with `NO_SHARING` because it implies "bundling" all files into the folder, including those potentially re-used after recovery.
Other strategies don't imply that.